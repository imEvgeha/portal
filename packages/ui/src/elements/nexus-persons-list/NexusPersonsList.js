/* eslint-disable */
import React, {useState, useEffect, useContext, useCallback} from 'react';
import PropTypes from 'prop-types';
import {useDispatch, useSelector} from 'react-redux';
import Button from '@atlaskit/button';
import UserPicker from '@atlaskit/user-picker';
import classnames from 'classnames';
import {DragDropContext, Droppable} from 'react-beautiful-dnd';
import {uid} from 'react-uid';
import {NexusModalContext} from '@vubiquity-nexus/portal-ui/lib/elements/nexus-modal/NexusModal';

import PropagateForm from '../../../../../src/pages/title-metadata/components/title-metadata-details/components/PropagateForm';
import {PROPAGATE_TITLE} from '../nexus-dynamic-form/constants';
import NexusPerson from '../nexus-person/NexusPerson';
import NexusPersonRO from '../nexus-person-ro/NexusPersonRO';
import {isObject} from '@vubiquity-nexus/portal-utils/lib/Common';
import {getDir} from '../nexus-dynamic-form/utils';
import {removeSeasonPerson} from '../../../../../src/pages/title-metadata/titleMetadataActions';
import {propagateRemovePersonsSelector} from '../../../../../src/pages/title-metadata/titleMetadataSelectors';
import CreateEditConfigForm from '../../../../../src/pages/legacy/containers/config/CreateEditConfigForm';
import {CAST, CAST_CONFIG, SEASON} from './constants';
import {loadOptions} from './utils';
import './NexusPersonsList.scss';
import {configService} from '../../../../../src/pages/legacy/containers/config/service/ConfigService';

const NexusPersonsList = ({
    personsList,
    uiConfig,
    getValues,
    setFieldValue,
    hasCharacter,
    isEdit,
    updateCastCrew,
    searchPerson,
    castCrewConfig,
    emetLanguage,
    setUpdate,
    isVerticalLayout,
    ...props
}) => {
    const dispatch = useDispatch();
    const {openModal, closeModal} = useContext(NexusModalContext);
    const [openPersonModal, setOpenPersonModal] = useState(false);
    const [currentRecord, setCurrentRecord] = useState({});
    const [persons, setPersons] = useState(personsList || []);
    const [searchText, setSearchText] = useState('');
    const propagateRemovePersons = useSelector(propagateRemovePersonsSelector);
    const {title, contentType, editorial, editorialMetadata} = getValues();

    useEffect(() => {
        const updatedPersons = [...personsList];
        updatedPersons.forEach((person, index) => {
            person.creditsOrder = index;
        });
        setPersons(updatedPersons);
    }, [personsList]);

    const searchInputChanged = val => {
        setSearchText(val);
        const input = document.getElementById(uiConfig.htmlFor);
        input.setAttribute('dir', getDir(val));
    };

    const isPersonValid = entry => {
        return (
            persons === null ||
            persons.findIndex(
                person =>
                    person.displayName.toString().toLowerCase() === entry.displayName.toString().toLowerCase() &&
                    person.personType.toString().toLowerCase() === entry.personType.toString().toLowerCase()
            ) < 0
        );
    };

    const validateAndAddPerson = personJSON => {
        if (isObject(personJSON) && personJSON.id === 'create') {
            setOpenPersonModal(true);
        } else {
            const person = JSON.parse(personJSON.original);
            const isValid = isPersonValid(person);
            if (isValid) {
                addPerson(person);
                setSearchText('');
            } else {
                openModal(
                    <Button appearance="primary" onClick={closeModal}>
                        OK
                    </Button>,
                    {
                        title: (
                            <div className="nexus-c-nexus-persons-list__error-modal-title">Person already exists!</div>
                        ),
                        width: 'small',
                    }
                );
            }
        }
    };

    const addPerson = person => {
        let updatedPersons = [...persons];

        if (person['personTypes'] && Array.isArray(person['personTypes'])) {
            let personWithType = {...person};
            delete personWithType['personTypes'];

            person['personTypes'].map(personType => {
                updatedPersons.push({...personWithType, personType: personType});
            });
        } else {
            updatedPersons.push(person);
        }

        const isCast = uiConfig.type === CAST;
        updatedPersons.forEach((person, index) => {
            person.creditsOrder = index;
        });
        setPersons(updatedPersons);
        updateCastCrew(updatedPersons, isCast);
    };

    const removePerson = person => {
        const updatedPersons = persons.filter(entry => {
            return entry.id !== person.id || entry.personType !== person.personType;
        });

        const isCast = uiConfig.type === CAST;
        updatedPersons.forEach((person, index) => {
            person.creditsOrder = index;
        });
        setPersons(updatedPersons);
        updateCastCrew(updatedPersons, isCast);

        if (!isVerticalLayout && contentType === SEASON) {
            let isDuplicate = false;
            propagateRemovePersons.forEach(entry => {
                if (entry.id === person.id && entry.personType === person.personType) {
                    isDuplicate = true;
                }
            });

            const {id, personType, creditsOrder} = person;
            const payload = isDuplicate
                ? propagateRemovePersons
                : [...propagateRemovePersons, {id, personType, creditsOrder}];

            dispatch(removeSeasonPerson(payload));
        }

        const updateEditorialMetadata = editorialMetadata.map(emet => {
            const updatedCastCrew =
                emet?.castCrew &&
                emet.castCrew.filter(entry => {
                    return entry.id !== person.id || entry.personType !== person.personType;
                });

            const updatedEmet = {
                ...emet,
                castCrew: updatedCastCrew,
            };

            if (updatedEmet.language === editorial.language && updatedEmet.locale === editorial.locale) {
                setFieldValue('editorial', updatedEmet);

                if (isVerticalLayout) {
                    return updatedEmet
                }
            }

            if (!isVerticalLayout) {
                return updatedEmet; 
            } else {
                return emet
            }
        });

        setFieldValue('editorialMetadata', updateEditorialMetadata);

        closeModal();
        setUpdate(prev => !prev);
    };

    const closePropagateModal = () => {
        closeModal();
        setUpdate(prev => !prev);
    };

    const openRemoveModal = person => {
        const removeMessage = () => {
            if (isVerticalLayout) {
                return `Remove ${person.displayName} from this Emet`;
            } else {
                if (contentType === SEASON) {
                    return `Remove ${person.displayName} from this Season, it's Episodes and all related Emets?`;
                } else {
                    return `Remove ${person.displayName} from ${title}?`;
                }
            }
        };

        openModal(
            <>
                <p>{removeMessage()}</p>
                <div className="nexus-c-nexus-persons-list__remove-modal-actions">
                    <Button onClick={closeModal}>Cancel</Button>
                    <Button onClick={() => removePerson(person)}>{'Remove'}</Button>
                </div>
            </>,
            {
                title: 'Remove',
                width: 'small',
            }
        );
    };

    const openPropagateModal = useCallback(person => {
        openModal(
            <PropagateForm
                person={person}
                getValues={getValues}
                setFieldValue={setFieldValue}
                onClose={closePropagateModal}
            />,
            {
                title: PROPAGATE_TITLE,
                width: 'small',
            }
        );
    }, []);

    const reorder = (list, startIndex, endIndex) => {
        const updatedPersons = [...list];
        const [removed] = updatedPersons.splice(startIndex, 1);
        updatedPersons.splice(endIndex, 0, removed);
        return updatedPersons;
    };

    const onDragEnd = result => {
        if (!result.destination) {
            return;
        }
        const updatedPersons = reorder(persons, result.source.index, result.destination.index);
        updatedPersons.forEach((person, index) => {
            person.creditsOrder = index;
        });
        setPersons(updatedPersons);
        const isCast = uiConfig.type === CAST;
        updateCastCrew(updatedPersons, isCast);
    };

    const onEditPerson = async person => {
        const endpoint = castCrewConfig && castCrewConfig.urls && castCrewConfig.urls['CRUD'];
        const personData = await configService.get(endpoint, person.id);
        setCurrentRecord(personData);
        setOpenPersonModal(true);
    };

    const makeDraggableContainer = content => {
        return (
            <DragDropContext onDragEnd={onDragEnd}>
                <Droppable droppableId="droppable">
                    {(provided, snapshot) => (
                        <div
                            ref={provided.innerRef}
                            {...provided.droppableProps}
                            className={classnames('nexus-c-nexus-persons-list__droppable-content', {
                                'nexus-c-nexus-persons-list__droppable-content--dragging': snapshot.isDraggingOver,
                            })}
                        >
                            {content}
                            {provided.placeholder}
                        </div>
                    )}
                </Droppable>
            </DragDropContext>
        );
    };

    const renderPersons = () => {
        return persons.map((person, i) => {
            const customKey = person.id ? uid(person.id, i) : `${person.displayName}-${i}`;
            return (
                <NexusPerson
                    key={customKey}
                    person={person}
                    customKey={customKey}
                    index={i}
                    hasCharacter={hasCharacter}
                    onRemove={() => openRemoveModal(person)}
                    onPropagate={() => openPropagateModal(person)}
                    onEditPerson={() => onEditPerson(person)}
                    emetLanguage={emetLanguage}
                    {...props}
                />
            );
        });
    };

    const renderPersonsRO = () => {
        return persons.map((person, i) => {
            return <NexusPersonRO key={uid(person.id, i)} person={person} emetLanguage={emetLanguage} />;
        });
    };

    const editRecord = val => {
        const newVal = {...currentRecord, ...val};
        const endpoint = castCrewConfig && castCrewConfig.urls && castCrewConfig.urls['CRUD'];
        if (newVal.id) {
            configService.update(endpoint, newVal.id, newVal).then(response => {
                const updatedList = persons.map(person => {
                    if (person.id === newVal.id) {
                        return {
                            displayName: response.displayName,
                            creditsOrder: person.creditsOrder,
                            personType: person.personType,
                            id: response.id,
                            firstName: response.firstName,
                            lastName: response.lastName,
                        };
                    }
                    return person;
                });
                setPersons(updatedList);
                setCurrentRecord({});
                updateCastCrew(updatedList, uiConfig.type === CAST);
                setOpenPersonModal(false);
            });
        } else {
            configService.create(endpoint, newVal).then(person => {
                setCurrentRecord({});
                setOpenPersonModal(false);
                addPerson(person);
                setSearchText('');
            });
        }
    };

    const closePersonModal = () => {
        setOpenPersonModal(false);
        setCurrentRecord({});
    };

    return (
        <>
            <div className="nexus-c-nexus-persons-list__heading">{uiConfig.title}</div>
            {isEdit ? (
                <>
                    {castCrewConfig && openPersonModal && (
                        <CreateEditConfigForm
                            onRemoveItem={() => {}}
                            schema={castCrewConfig && castCrewConfig.uiSchema}
                            label={castCrewConfig && castCrewConfig.displayName}
                            displayName={castCrewConfig && castCrewConfig.displayName}
                            value={currentRecord}
                            onSubmit={editRecord}
                            onCancel={closePersonModal}
                        />
                    )}
                    <div className="nexus-c-nexus-persons-list__add">
                        <UserPicker
                            fieldId={uiConfig.htmlFor}
                            inputId={uiConfig.htmlFor}
                            width="100%"
                            loadOptions={() => loadOptions(uiConfig, searchText, searchPerson, emetLanguage)}
                            value={searchText}
                            onInputChange={searchInputChanged}
                            onSelection={validateAndAddPerson}
                            placeholder={uiConfig.newLabel}
                        />
                    </div>
                    {makeDraggableContainer(persons && renderPersons())}
                </>
            ) : (
                <>{persons && renderPersonsRO()}</>
            )}
        </>
    );
};

NexusPersonsList.propTypes = {
    onChange: PropTypes.func,
    personsList: PropTypes.array,
    uiConfig: PropTypes.object,
    hasCharacter: PropTypes.bool,
    isEdit: PropTypes.bool,
    updateCastCrew: PropTypes.func,
    setFieldValue: PropTypes.func,
    getValues: PropTypes.func,
    searchPerson: PropTypes.func,
    castCrewConfig: PropTypes.object,
    emetLanguage: PropTypes.string,
    setUpdate: PropTypes.func,
    isVerticalLayout: PropTypes.bool,
};

NexusPersonsList.defaultProps = {
    personsList: [],
    uiConfig: CAST_CONFIG,
    hasCharacter: false,
    isEdit: false,
    updateCastCrew: () => null,
    setFieldValue: () => null,
    getValues: () => null,
    searchPerson: () => null,
    setUpdate: () => null,
    castCrewConfig: {},
    emetLanguage: 'en',
    isVerticalLayout: false,
};

export default NexusPersonsList;
