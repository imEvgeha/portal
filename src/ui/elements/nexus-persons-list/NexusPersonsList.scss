.nexus-c-persons-list {
    min-height: 160px;
}

.nexus-c-persons-list__group {
    overflow-x: hidden;
}

.nexus-c-persons-list__add {
    margin-top: 5px;
    border: none;
    border-radius: 3px;
    width: 97%;

    &.invalid {
        border: 2px solid red;
    }
}

.nexus-c-persons-list__add-error {
    color: #e74c3c;
    font-weight: bold;
}
