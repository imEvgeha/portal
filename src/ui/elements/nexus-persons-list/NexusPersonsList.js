/* eslint-disable */
import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {cloneDeep} from 'lodash';
import {Label} from '@atlaskit/field-base';
import UserPicker from '@atlaskit/user-picker';
import {DragDropContext, Droppable} from 'react-beautiful-dnd';
import {uid} from 'react-uid';
import {ListGroup, Card, CardHeader, CardBody} from 'reactstrap';
import {getFilteredCastList, getFilteredCrewList} from '../../../pages/legacy/constants/metadata/configAPI';
import DroppableContent from './elements/DroppableContent';
import {searchPerson} from './utils';
import NexusCharacterNameModal from './elements/NexusCharacterNameModal';
import NexusPerson from './elements/NexusPerson';
import NexusPersonRO from './elements/NexusPersonRO';
import {CAST, CAST_CONFIG, PERSONS_PER_REQUEST} from './constants';
import './NexusPersonsList.scss';

const NexusPersonsList = ({personsList, uiConfig, hasCharacter, showPersonType, isEdit, updateCastCrew}) => {
    const [persons, setPersons] = useState(personsList || []);
    const [isLastEntryValid, setLastIsEntryValid] = useState(true);
    const [searchText, setSearchText] = useState('');
    const [modalData, setModalData] = useState({});
    const [isModalOpen, setIsModalOpen] = useState(false);
    const [selectedPersonIndex, setSelectedPersonIndex] = useState(null);
    const [modalInputValue, setModalInputValue] = useState('');

    const searchInputChanged = val => {
        setSearchText(val);
    };

    const loadOptions = () => {
        const {type} = uiConfig;
        if (searchText.length < 2) return [];
        if (type === CAST) {
            return searchPerson(searchText, PERSONS_PER_REQUEST, type, true).then(res =>
                getFilteredCastList(res.data, true, true).map(e => {
                    return {
                        id: e.id,
                        name: e.displayName,
                        byline: e.personType.toString().toUpperCase(),
                        original: JSON.stringify(e),
                    };
                })
            );
        } else {
            return searchPerson(searchText, PERSONS_PER_REQUEST, type).then(res =>
                getFilteredCrewList(res.data, true).map(e => {
                    return {
                        id: e.id,
                        name: e.displayName,
                        byline: e.personType.toString().toUpperCase(),
                        original: JSON.stringify(e),
                    };
                })
            );
        }
    };

    const isPersonValid = entry => {
        return (
            persons === null ||
            persons.findIndex(
                person =>
                    person.id === entry.id &&
                    person.personType.toString().toLowerCase() === entry.personType.toString().toLowerCase()
            ) < 0
        );
    };

    const validateAndAddPerson = personJSON => {
        const person = JSON.parse(personJSON.original);
        const isValid = isPersonValid(person);
        if (isValid) {
            addPerson(person);
            setSearchText('');
        }
        setLastIsEntryValid(isValid);
    };

    const addPerson = person => {
        const updatedPersons = [...[person], ...persons];
        const isCast = uiConfig.type === CAST;
        setPersons(updatedPersons);
        updateCastCrew(updatedPersons, isCast);
    };

    const removePerson = person => {
        const updatedPersons = persons.filter(entry => {
            return entry.id !== person.id;
        });
        const isCast = uiConfig.type === CAST;
        setPersons(updatedPersons);
        updateCastCrew(updatedPersons, isCast);
    };

    const onCharacterSubmit = () => {
        const updatedPersons = cloneDeep(persons);
        const isCast = uiConfig.type === CAST;
        updatedPersons[selectedPersonIndex].characterName = modalInputValue;
        setPersons(updatedPersons);
        setSelectedPersonIndex(null);
        setModalInputValue('');
        updateCastCrew(updatedPersons, isCast);
        closeModal();
    };

    const closeModal = () => {
        setModalData({});
        setIsModalOpen(false);
        setSelectedPersonIndex(null);
        setModalInputValue('');
    };

    const openModal = id => {
        setSelectedPersonIndex(id);
        const selectedPerson = persons && persons[id];
        setModalInputValue(selectedPerson.characterName);
        setModalData({
            characterName: selectedPerson.characterName,
            displayName: selectedPerson.displayName,
        });
        setIsModalOpen(true);
    };

    const onModalInputChanged = event => {
        const {value} = event.target;
        setModalInputValue(value);
    };

    const reorder = (list, startIndex, endIndex) => {
        const result = Array.from(list);
        const [removed] = result.splice(startIndex, 1);
        result.splice(endIndex, 0, removed);

        return result;
    };

    const onDragEnd = result => {
        if (!result.destination) {
            return;
        }
        setPersons(reorder(persons, result.source.index, result.destination.index));
    };

    const makeDraggableContainer = content => {
        return (
            <DragDropContext onDragEnd={onDragEnd}>
                <Droppable droppableId="droppable">
                    {(provided, snapshot) => (
                        <div ref={provided.innerRef} {...provided.droppableProps}>
                            <DroppableContent isDragging={snapshot.isDraggingOver}>
                                {content}
                                {provided.placeholder}
                            </DroppableContent>
                        </div>
                    )}
                </Droppable>
            </DragDropContext>
        );
    };

    return (
        <>
            <Card className="nexus-c-persons-list">
                <CardHeader className="clearfix">
                    <h4 className="float-left">{uiConfig.title}</h4>
                </CardHeader>
                <CardBody>
                    <ListGroup className="nexus-c-persons-list__group">
                        <Label label={isEdit && uiConfig.newLabel} isFirstChild htmlFor={uiConfig.htmlFor}>
                            {isEdit && (
                                <div className={`nexus-c-persons-list__add ${isLastEntryValid ? '' : 'invalid'}`}>
                                    <UserPicker
                                        fieldId={uiConfig.htmlFor}
                                        width="100%"
                                        loadOptions={loadOptions}
                                        value={searchText}
                                        onInputChange={searchInputChanged}
                                        onSelection={validateAndAddPerson}
                                        placeholder={uiConfig.newLabel}
                                    />
                                </div>
                            )}
                            {isEdit && !isLastEntryValid && (
                                <span className="nexus-c-persons-list__add-error">Person already exists!</span>
                            )}
                            <Label label={isEdit && uiConfig.listLabel} isFirstChild>
                                {isEdit
                                    ? makeDraggableContainer(
                                          persons &&
                                              persons.map((person, i) => {
                                                  return (
                                                      <NexusPerson
                                                          key={uid(person.id, i)}
                                                          person={person}
                                                          index={i}
                                                          hasCharacter={hasCharacter}
                                                          showPersonType={showPersonType}
                                                          onRemove={() => removePerson(person)}
                                                          onEditCharacter={index => openModal(index)}
                                                      />
                                                  );
                                              })
                                      )
                                    : persons &&
                                      persons.map((person, i) => {
                                          return (
                                              <NexusPersonRO
                                                  key={uid(person.id, i)}
                                                  person={person}
                                                  showPersonType={showPersonType}
                                              />
                                          );
                                      })}
                            </Label>
                        </Label>
                    </ListGroup>
                </CardBody>
            </Card>
            <NexusCharacterNameModal
                onSubmit={onCharacterSubmit}
                displayName={modalData.displayName}
                characterName={modalData.characterName}
                value={modalInputValue}
                onChange={onModalInputChanged}
                isModalOpen={isModalOpen}
                closeModal={closeModal}
            />
        </>
    );
};

NexusPersonsList.propTypes = {
    onChange: PropTypes.func,
    personsList: PropTypes.array,
    uiConfig: PropTypes.object,
    hasCharacter: PropTypes.bool,
    showPersonType: PropTypes.bool,
    isEdit: PropTypes.bool,
    updateCastCrew: PropTypes.func,
};

NexusPersonsList.defaultProps = {
    personsList: [],
    uiConfig: CAST_CONFIG,
    hasCharacter: false,
    showPersonType: true,
    isEdit: false,
    updateCastCrew: () => null,
};

export default NexusPersonsList;
