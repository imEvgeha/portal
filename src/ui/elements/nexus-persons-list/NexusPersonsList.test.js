import React from 'react';
import {shallow} from 'enzyme';
import NexusPersonsList from './NexusPersonsList';

describe('NexusPersonsList', () => {
    let wrapper;
    const personsList = [
        {
            id: 'prs_Ac9oyaREip',
            firstName: null,
            middleName: null,
            lastName: null,
            displayName: 'Amaka Obi',
            personType: 'Actor',
            characterName: null,
        },
        {
            id: 'prs_airo586ETR',
            firstName: null,
            middleName: null,
            lastName: null,
            displayName: 'Ebube Nwagbo',
            personType: 'Actor',
            characterName: null,
        },
        {
            id: 'prs_x9TX5AkETo',
            firstName: null,
            middleName: null,
            lastName: null,
            displayName: 'Solomon Akiyesi',
            personType: 'Actor',
            characterName: null,
        },
    ];
    it('should match snapshot', () => {
        wrapper = shallow(
            <NexusPersonsList
                personsList={personsList}
                hasCharacter
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});

