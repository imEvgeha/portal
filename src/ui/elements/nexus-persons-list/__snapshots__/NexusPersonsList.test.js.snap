// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NexusPersonsList should match snapshot 1`] = `
<Fragment>
  <div
    className="nexus-c-persons-list__heading"
  >
    Cast
  </div>
  <NexusPersonRO
    key="val-prs_Ac9oyaREip"
    person={
      Object {
        "characterName": null,
        "displayName": "Amaka Obi",
        "firstName": null,
        "id": "prs_Ac9oyaREip",
        "lastName": null,
        "middleName": null,
        "personType": "Actor",
      }
    }
  />
  <NexusPersonRO
    key="idx-1"
    person={
      Object {
        "characterName": null,
        "displayName": "Ebube Nwagbo",
        "firstName": null,
        "id": "prs_airo586ETR",
        "lastName": null,
        "middleName": null,
        "personType": "Actor",
      }
    }
  />
  <NexusPersonRO
    key="idx-2"
    person={
      Object {
        "characterName": null,
        "displayName": "Solomon Akiyesi",
        "firstName": null,
        "id": "prs_x9TX5AkETo",
        "lastName": null,
        "middleName": null,
        "personType": "Actor",
      }
    }
  />
</Fragment>
`;
